### ISSUE TITLE
Running userscript from Toolbox splits focus with Toolbox "Filter Tools" field

### ISSUE THREAD

**Issue by mjross** [Komodo/KomodoEdit#2702]
### Summary
Running a userscript from its Toolbox context menu splits Komodo IDE's focus between the userscript (where it belongs) and the Toolbox "Filter Tools" field (where it doesn't). Consequently, when he then types a key sequence, that action occurs not only within the userscript but also in the Toolbox, which probably was not the intended result.

### Details
When a user is editing a userscript, he typically will want to test the changes frequently, by right clicking on the name of the userscript in the Toolbox within the right pane, and clicking "Execute Userscript". After the userscript is run, he will usually want to make further modifications to it or look at another open file (usually by employing a key sequence to go to that file, e.g., Ctrl+PageDown). But when executing a userscript, focus is seemingly split between the userscript itself and the "Filter Tools" search field, located at the top of the Toolbox, even though the user probably had no intention of using this field. (After all, the userscript he wanted to execute was visible and then was executed. Most likely he would continue working on that script, and not try to locate a different script using that search field.)

As a consequence, after running the userscript, when he types a key sequence to continue his editing, that action occurs not only within the userscript but also in the Toolbox. For instance, if he types Ctrl+PageDown, then Komodo IDE displays the next open file (if any) – as it should – but also changes the right pane so focus in the right pane is no longer on the Toolbox search field but instead on the Code widget. That second action should not take place; the user most likely did not intend to go to the Code widget, and it forces him to go back to the Toolbox widget each time. Essentially, the edit cursor should not be placed in the Toolbox search field automatically just because the user utilized a userscript's context menu to run it.

### Platform Information
Komodo IDE 10.2.2 on Windows 10 Pro

### Attachments
![code widget now has focus](https://user-images.githubusercontent.com/829492/28491739-6e4ecea0-6eb3-11e7-9e48-a8e30b921838.png)

**Comment by Naatan** [comment]
Can you share a userscript that reproduces this?

**Comment by mjross** [comment]
Certainly. The (mis)behavior is the same, regardless of the code in the userscript that is run. The first screenshot shows Komodo IDE just before I press Enter (to dismiss the alert) and then Ctrl+PageDown. The second screenshot shows what it looks like after; you can see that in the edit pane it has changed to a different file (correctly), but in the right pane it too has changed, from the Toolbox widget to a different one (incorrectly). (Sorry if my terminology is inaccurate.) So it's as if running the userscript using the context menu item has divided the focus of Komodo IDE so that the next key sequence affects two different parts of the interface.

![komodo ide after running userscript](https://user-images.githubusercontent.com/829492/28534709-cf6bccfe-705e-11e7-8b3b-34e7577a0390.jpg)

![komodo ide after ctrl pagedown](https://user-images.githubusercontent.com/829492/28534720-d8684ea4-705e-11e7-8275-d730b007955c.jpg)

**Comment by Naatan** [comment]
Ok, from what I'm seeing when I test this the ctrl+pagedown is only ran on the side panel, not on the editor. Which is by design because that is what had focus when the userscript was executed. Are you saying you are seeing your keystrokes get executed on the editor too?

**Comment by mjross** [comment]
Yes, the keystrokes are executed on the editor as well.

But I would argue that that is the correct and only place where the
keystrokes should be executed, and not in the Toolbox. Now if a user were
to left-click on anything in the Toolbox, especially the search field, then
it should then receive focus. But in this case, right-clicking on the
userscript name should not transfer focus to the Toolbox. In almost every
UI that I've seen, the general rule is that a left-click does change focus
but a right-click does not. In this instance, the user is only
right-clicking on something in the Toolbox because that is where the
userscript name is displayed, but he is still interested in the userscript
code itself (not the Toolbox per se), and so the focus should stay there.

**Comment by Naatan** [comment]
I'm unable to reproduce this on either Linux or Windows. @cgchoffman @mjainQA can you try?

@mjross you seem to have skipped the steps to reproduce fields when filing, please be sure to include those in the future, they are vital :)

**Comment by Naatan** [comment]
@mjross added steps to reproduce to your initial report, please verify them.

**Comment by mjross** [comment]
Sorry. Thanks for adding those steps.

I changed the expected results from "The side panel tab changes" to "The editor panel tab changes".

**Comment by Naatan** [comment]
The expected result is for the side panel tab to change, unless the userscript itself shifts focus. The side panel is what had focus when it was executed, as such it makes sense that ctrl+page down is executed against the panel, not the editor. It's up to the developer to manage this logic as needed, for us to change it so eg. it is always ran against the editor would be breaking at this point.

**Comment by mjross** [comment]
There seem to be two separate but related issues: 1) The behavior of the
key sequence affecting both the editor and side panes, i.e., both of them
have focus. I understand that you guys are not seeing that behavior, but
I'm able to reproduce it every time. I'm not sure if this indicates some
type of corruption within my Komodo IDE installation. I'm willing to do a
clean reinstall, if needed.

2) The question as to which pane should have focus after the user right
clicks on the userscript's name in the Toolbox. I've already stated my
reasons as to why the focus should remain with the edit pane and not be
transferred to the search field of the Toolbox. If that's a decision for
others to make, no problem. But simply accessing the context menu of a
userscript, in my opinion, should not transfer focus to an unrelated search
field, and away from the userscript that the user is executing and editing.

**Comment by Naatan** [comment]
We are addressing the first issue here. The second issue is a non-issue as this logic is owned by the script. I would agree that it would have been better for this to have defaulted its focus to the editor, but this is not how it was implemented and changing it now would likely break a ton of userscripts. 

In other words it is not a bug as it was by design. Whether we agree with that design or not is at this point mostly irrelevant as changing the design would break userscripts that adopted this behaviour.

**Comment by mjross** [comment]
Okay. Just let me know if you want me to try anything to help resolve the first issue.

As for the second issue, I had no idea that userscripts could work with items within any of the widgets. Man, I have a lot to learn about them. (Related to that, I have a question about them, but won't post it here but will instead post it to the Support forum.)

**Comment by Naatan** [comment]
They can work with anything an addon can. Skies the limit.

**Comment by mjross** [comment]
Cool. It sounds like they have a lot of potential.
